File: tests/fixtures/swift/Service.swift
import Foundation

// User struct
struct User {
    let id: Int
    let name: String
    let email: String
}

// Repository protocol
protocol Repository {
    func findById(_ id: Int) -> User?
    func save(_ user: User)
    func count() -> Int
}

// Logger protocol
protocol Logger {
    func log(_ message: String)
}

// UserService class
class UserService {
    private let repo: Repository
    private let logger: Logger

    init(repo: Repository, logger: Logger) {
        self.repo = repo
        self.logger = logger
    }

    func findById(_ id: Int) -> User? {
        logger.log("Finding user \(id)")
        return repo.findById(id)
    }

    func save(_ user: User) {
        repo.save(user)
    }

    func count() -> Int {
        return repo.count()
    }
}

// AbstractController class
class AbstractController {
    let viewPath: String

    init(viewPath: String) {
        self.viewPath = viewPath
    }

    func render(_ template: String) -> String {
        return "\(viewPath)/\(template)"
    }
}

// UserController extends AbstractController
class UserController: AbstractController {
    private let userService: UserService

    init(userService: UserService) {
        self.userService = userService
        super.init(viewPath: "users")
    }

    func index() -> String { return render("index") }
    func show(_ id: Int) -> String {
        let user = userService.findById(id)
        return render("show")
    }
}

// Utils enum (namespace)
enum Utils {
    static func formatName(_ name: String) -> String {
        return "User: \(name)"
    }
}

File: tests/fixtures/ruby/service.rb
# User class
class User
  attr_accessor :id, :name, :email

  def initialize(id, name, email)
    @id = id
    @name = name
    @email = email
  end
end

# Logger module
module Logger
  def log(message)
    puts "[LOG] #{message}"
  end
end

# Repository module (interface)
module Repository
  def find_by_id(id); end
  def save(user); end
  def count; end
end

# UserService class
class UserService
  include Logger

  def initialize(repo, logger = nil)
    @repo = repo
    @logger = logger
  end

  def find_by_id(id)
    @repo.find_by_id(id)
  end

  def save(user)
    @repo.save(user)
  end

  def count
    @repo.count
  end
end

# AbstractController class
class AbstractController
  attr_reader :view_path

  def initialize(view_path)
    @view_path = view_path
  end

  def render(template)
    "#{view_path}/#{template}"
  end
end

# UserController extends AbstractController
class UserController < AbstractController
  def initialize(user_service)
    super("users")
    @user_service = user_service
  end

  def index
    render("index")
  end

  def show(id)
    user = @user_service.find_by_id(id)
    render("show")
  end
end
